Dummy PR only here for the purpose of review

sloriot opened the pull request.

kkatrio committed:
is_degenerate_edge function

kkatrio committed:
is_degenerate_triangle_face function

kkatrio committed:
duplicate_vertices function doc

kkatrio committed:
is_non_manifold_vertex function

kkatrio committed:
is needle andcap functions

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+  }while(h != start);
+
+  BOOST_FOREACH(halfedge_descriptor h, halfedges_around_target(v, tm))
+  {
```
this loop is exactly the one above.
The first loop needs to be on all halfedges of the mesh and you should add all halfedges having `v` as target in `halfedges_handled`.

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+
+/// \ingroup PMP_repairing_grp
+/// checks whether an edge is degenerate.
+/// An edge is considered degenerate if two of its vertices share the same location.
```
if the points of its vertices are identical

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+/// checks whether an edge is degenerate.
+/// An edge is considered degenerate if two of its vertices share the same location.
+///
+/// @tparam PolygonMesh a model of `FaceListGraph` and `MutableFaceGraph`
```
`HalfedgeGraph` is sufficient

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+/// @tparam NamedParameters a sequence of \ref pmp_namedparameters "Named Parameters"
+///
+/// @param e the edge to check whether is degenerate
+/// @param pm the triangulated surface mesh upon evaluation
```
no needs to be triangulated

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+/// @param tm the triangulated surface mesh upon evaluation
+///
+/// \return true if the vertrex is non-manifold
+template <typename TriangleMesh>
```
TriangleMesh ->PolygonMesh. The algorithm is not only working on triangle meshes

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+/// checks whether a triangle face is degenerate.
+/// A triangle face is considered degenerate if all three points of the face are collinear.
+///
+/// @tparam TriangleMesh a model of `FaceListGraph` and `MutableFaceGraph`
```
`FaceGraph` is sufficient

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+
+/// \ingroup PMP_repairing_grp
+/// checks whether a triangle face is degenerate.
+/// A triangle face is considered degenerate if all three points of the face are collinear.
```
A triangle is degenerate if its there points are collinear.

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+/// \cgalNamedParamsBegin
+///    \cgalParamBegin{vertex_point_map} the property map with the points associated to the vertices of `pmesh`. The type of this map is model of `ReadWritePropertyMap`.
+/// If this parameter is omitted, an internal property map for
+/// `CGAL::vertex_point_t` should be available in `PolygonMesh`
```
PolygonMesh -> TriangleMesh

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -732,7 +1013,7 @@ std::size_t remove_degenerate_faces(TriangleMesh& tmesh,
   std::set<face_descriptor> degenerate_face_set;
   BOOST_FOREACH(face_descriptor fd, faces(tmesh))
-    if ( is_degenerate_triangle_face(fd, tmesh, vpmap, traits) )
+    if ( is_degenerate_triangle_face(fd, tmesh) )
```
pass `np` to the function

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -763,7 +1044,7 @@ std::size_t remove_degenerate_faces(TriangleMesh& tmesh,
       // remove the central vertex and check if the new face is degenerated
       hd=CGAL::Euler::remove_center_vertex(hd, tmesh);
-      if (is_degenerate_triangle_face(face(hd, tmesh), tmesh, vpmap, traits))
+      if (is_degenerate_triangle_face(face(hd, tmesh), tmesh))
```
pass `np` to the function

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+  Traits traits = choose_param(get_param(np, internal_np::geom_traits), Traits());
+
+  // call from BGL helpers
+  return is_degenerate_triangle_face(f, tm, vpmap, traits);
```
remove the implementation from BGL and inline it here. Note you'll have to update a couple of files that were using the old function. General remark, the following `const typename Traits::Point_3& p1 = get(vpm, v)` is not correct, it should be `typename boost::property_traits<Vpm>::reference p1 = get(vpm, v)`.

sloriot replied:
inlining is missing

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+}
+
+/// \ingroup PMP_repairing_grp
+/// checks whether a triangle face is needle-like.
```
is a needle.

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+
+/// \ingroup PMP_repairing_grp
+/// checks whether a triangle face is needle-like.
+/// In a needle-like triangle its longest edge is much longer than the shortest one.
```
A triangle is a needle if ...

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+/// \cgalNamedParamsBegin
+///    \cgalParamBegin{vertex_point_map} the property map with the points associated to the vertices of `pmesh`. The type of this map is model of `ReadWritePropertyMap`.
+/// If this parameter is omitted, an internal property map for
+/// `CGAL::vertex_point_t` should be available in `PolygonMesh`
```
TriangleMesh

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+
+    Vector a = get(vpmap, v0) - get(vpmap, v1);
+    Vector b = get(vpmap, v2) - get(vpmap, v1);
+    double ab = a*b;
```
`GT::FT`

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+    double ab = a*b;
+    double aa = a.squared_length();
+    double bb = b.squared_length();
+    double dot_ab = a*b / (CGAL::sqrt(aa) * CGAL::sqrt(bb));
```
avoid using sqrt

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+    vertex_descriptor v2 = target(next(h, tm), tm);
+
+    Vector a = get(vpmap, v0) - get(vpmap, v1);
+    Vector b = get(vpmap, v2) - get(vpmap, v1);
```
do no construct vectors, use `CGAL:squared_distance()` or `GT::Compute_squared_distance()`

kkatrio replied:
but then how will I calculate the dot product of the euclidean vectors? the objective is to calculate the cosine of their angle

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+    double dot_ab = a*b / (CGAL::sqrt(aa) * CGAL::sqrt(bb));
+
+    // threshold = 1 means no tolerance, totally degenerate
+    if(dot_ab > threshold)
```
what about `bc`?

kkatrio replied:
what is bc

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+/// @param threshold a number in the range [0, 1] to indicate the tolerance
+///        upon which to characterize the degeneracy. 1 means that needle triangles
+///        are those that have a infinitely small edge, while 0 means that needle triangles
+///        are those that would have an infinitely long edge
```
I don't understand what you mean here. Wouldn't it be easier to use the ratio between the larger and the smallest edge?

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+}
+
+/// \ingroup PMP_repairing_grp
+/// checks whether a triangle face is cap-like.
```
is a cap.

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+
+/// \ingroup PMP_repairing_grp
+/// checks whether a triangle face is cap-like.
+/// A cap-like triangle has an angle very close to 180 degrees.
```
A triangle is a cap if ...

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+/// checks whether a triangle face is needle-like.
+/// In a needle-like triangle its longest edge is much longer than the shortest one.
+///
+/// @tparam TriangleMesh a model of `FaceListGraph` and `MutableFaceGraph`
```
FaceGraph is sufficient

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+/// checks whether a triangle face is cap-like.
+/// A cap-like triangle has an angle very close to 180 degrees.
+///
+/// @tparam TriangleMesh a model of `FaceListGraph` and `MutableFaceGraph`
```
FaceGraph is sufficient

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+/// \cgalNamedParamsBegin
+///    \cgalParamBegin{vertex_point_map} the property map with the points associated to the vertices of `pmesh`. The type of this map is model of `ReadWritePropertyMap`.
+/// If this parameter is omitted, an internal property map for
+/// `CGAL::vertex_point_t` should be available in `PolygonMesh`
```
TriangleMesh + tm

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+/// @param threshold a number in the range [0, 1] to indicate the tolerance
+///        upon which to characterize the degeneracy. 1 means that cap triangles
+///        are considered those whose vertices for an angle of 180 degrees, while 0 means that
+///        all triangles are considered caps.
```
I guess it would be more easy to understand if it is a bound on the cosinus of the largest angle in the triangle.

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+/// @tparam TriangleMesh a model of `FaceListGraph` and `MutableFaceGraph`
+/// @tparam NamedParameters a sequence of \ref pmp_namedparameters "Named Parameters"
+///
+/// @param f the face to check whether is almost degenerate
```
a face

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+/// @tparam NamedParameters a sequence of \ref pmp_namedparameters "Named Parameters"
+///
+/// @param f the face to check whether is almost degenerate
+/// @param tm the triangulated surface mesh upon evaluation
```
triangle mesh containing `f` same in the other function.

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+    double ab = a*b;
+    double aa = a.squared_length();
+    double bb = b.squared_length();
+    double dot_ab = a*b / (CGAL::sqrt(aa) * CGAL::sqrt(bb));
```
try to avoid using sqrt

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -186,6 +184,289 @@ degenerate_faces(const TriangleMesh& tm, OutputIterator out)
+
+    Vector a = get(vpmap, v0) - get(vpmap, v1);
+    Vector b = get(vpmap, v2) - get(vpmap, v1);
+    double ab = a*b;
```
`GT::FT`

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -1281,17 +1562,45 @@ std::size_t remove_degenerate_faces(TriangleMesh& tmesh,
+/// @tparam NamedParameters a sequence of \ref pmp_namedparameters "Named Parameters"
+///
+/// @param tm the triangulated surface mesh to be repaired
+/// @param np optional \ref pmp_namedparameters "Named Parameters" described below
```
Are "" used only here or is it generally use? Italic should be used IMO

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -1281,17 +1562,45 @@ std::size_t remove_degenerate_faces(TriangleMesh& tmesh,
+/// \ingroup PMP_repairing_grp
+/// duplicates all non-manifold vertices of the input mesh.
+///
+/// @tparam TriangleMesh a model of `FaceListGraph` and `MutableFaceGraph`
```
HalfedgeListGraph and MutableHalfedgeGraph should be sufficient.

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair.h:
```diff
@@ -1281,17 +1562,45 @@ std::size_t remove_degenerate_faces(TriangleMesh& tmesh,
+///    \cgalParamEnd
+/// \cgalNamedParamsEnd
+///
+/// \return true if the triangle face is degenerate
```
I think it might be useful to be able to mark non-manifold vertices, if `vertex_is_constrained_map is passed` for example. Maybe also a way to get the correspondance between those vertices.  For example a new named parameter could take an output iterator where vector of vertex could be put (one vector per group of vertices that were corresponding to the same vertex).

sloriot commented on Polygon_mesh_processing/test/Polygon_mesh_processing/remove_degeneracies_test.cpp:
```diff
@@ -38,6 +137,12 @@ int main()
   fix("data_degeneracies/trihole.off");
+  check_edge_degeneracy("data_degeneracies/degtri_edge.off");
+  check_triangle_face_degeneracy("data_degeneracies/degtri_four.off");
+  test_vetices_duplication("data_degeneracies/degtri_four.off");
```
vetices

sloriot commented on Polygon_mesh_processing/test/Polygon_mesh_processing/remove_degeneracies_test.cpp:
```diff
@@ -38,6 +137,12 @@ int main()
+  test_vetices_duplication("data_degeneracies/degtri_four.off");
+  test_vertex_non_manifoldness("data/non_manifold_vertex.off");
+  test_needle("data_degeneracies/needle.off");
+  test_cap("data_degeneracies/cap.off");
```
your test should test not only the API but you should have assertions checking that some hand-made elements are or are not matching.

kkatrio committed:
add a couple of tests to cmakelists

kkatrio committed:
merge vertices, tests & data

sloriot committed:
rewrite boundary cycle merging

sloriot committed:
rename header and test file

sloriot committed:
add missing overload

sloriot committed:
add a function to merge vertices globally

sloriot committed:
remove incorrect optimisation

kkatrio committed:
corrections after the review

kkatrio committed:
Merge remote-tracking branch 'refs/remotes/kkatrio/repair_functions' into repair_functions

kkatrio committed:
move predicates to helper.h and seperate test file

kkatrio committed:
replace is_degenerate_triangle_face predicate with new version from PMP helpers

kkatrio committed:
use cosine for threshold on needles and caps

kkatrio committed:
named parameters for duplicate non-manifold vertices

sloriot committed:
WIP correctly linking halfedges around merged vertices ...
also disable the merge between cycles as it is not straight forward
it will be always possible

sloriot committed:
directly sort halfedges and use the ordering to detect illegal merges

a merge is considered as illegal if it makes to vertices
to be merged unreachable. For now if a cycle contain an illegal
merge, all merges of the cycle are ignored.

kkatrio committed:
add missing named parameter documentation

kkatrio committed:
documentation on merge border vertices functions
